### ISSUE TITLE
CTL-787: Enormous performance issues due ModelBase.Equals and ModelBase.GetHashCode

### ISSUE THREAD

**Issue by GeertvanHorrik** [Catel/Catel#776]
Jira issue originally created by user DanielKuehner:

I don't really know where to start and how to explain, so let me try.

Let's start with our situation: We want to save a tree with <sub>100.000 nodes to a XML (and some other stuff). For this the profiler tells us we need </sub>6450s. Looking at the picture shows us that <sub>6350s are spent in ModelBase.Equals and ModelBase.GetHashCode ... wow.

So I did some tests with setting an own EquialityComparer to my Dictionary simulating your functions (without property checks) and startet to alter them. My new test was smaller and took 46s to save the data.
1. I altered GetHashCode() to just
```
return obj.GetHashCode();
```
This reduced the saving time to 23s.
2. I altered Equals() to just
```
return ReferenceEquals(x, y)
```
This reduced the saving time to 7s - great thing!

Then I did some further step and overrided ModelBase.Equals and ModelBase.GetHashCode and placed following lines into them:
```
return RuntimeHelpers.Equals(this, obj);
```
and
```
return RuntimeHelpers.GetHashCode(this);
```
This reduced the saving time to 2s - yeah!! This is factor of 23!!! :-)

I believe the changes in ModelBase.GetHashCode of [CTL-178](https://catelproject.atlassian.net/browse/CTL-178) were worse doubling the times. I don't understand why you are doing a hashing on the always equal class name. This destroys every try to optimize code with hash sets because all instance fell into the same bucket and forcing an O( n ) search.

Enabling an optional equal check on the content of an instance is nice. Using an equality comparer is even nicer - but you need the type checks which slow down everything by the factor of </sub>3-4.
In other words: if you don't need this content check you still have an unwanted overhead which can massivly slow down an application - if you don't know about this.

At this point I would like to discuss this point the get a good solution for everybody.

The solution with re-overriding Equals and GetHashCode with having the original lines of codes works for me and is ok. But I don't know if I really like this.
Could you imagine a static flag to disable the whole usage of an Equality comparer instead of replacing it?
Or dismiss the whole code so that people who want to do a content comparison have to use the ModelEqualityComparer directly and in example give it to collections ... as inteded?

What do you think? Why does these lines of code have such an impact on the performance?

**Comment by GeertvanHorrik** [comment]
I am going to put them back to the default implementation and see how people respond to this in the beta of Catel v5.

**Comment by GeertvanHorrik** [comment]
Done.

**Comment by GeertvanHorrik** [comment]
Benchmarks will show up @ http://github.com/catel/catel.benchmarks

**Comment by DanielKuehner** [comment]
@GeertvanHorrik this will be big testing on my side ... thanks!
